File: cpp/src/analytic.cpp
#include "acpd/analytic.hpp"
#include <Eigen/SVD>
#include <Eigen/QR>
#include <algorithm>
#include <cmath>
namespace acpd {
    std::vector<std::array<int,3>> exponents(int d,int degree) {
        if((d!=2 && d!=3) || degree<0 || degree>10) throw std::invalid_argument("basis requires dimension 2/3 and degree 0..10");
        std::vector<std::array<int,3>> out;
        for(int r=0;r<=degree;++r) for(int a=r;a>=0;--a) {
            if(d==2) out.push_back({
                a,r-a,0
            });
            else for(int b=r-a;b>=0;--b) out.push_back({
                a,b,r-a-b
            });
        }
        return out;
    }
    Matrix basis(const Matrix& points,int degree) {
        validate_cloud(points,"basis points");
        const auto powers=exponents(static_cast<int>(points.cols()),degree);
        Matrix phi(points.rows(),static_cast<int>(powers.size()));
        for(int i=0;i<points.rows();++i) for(int k=0;k<phi.cols();++k) {
            double term=1;
            for(int a=0;a<points.cols();++a) for(int j=1;j<=powers[k][a];++j) term*=points(i,a)/j;
            phi(i,k)=term;
        }
        require_finite(phi,"Taylor basis");
        return phi;
    }
    std::vector<int> degree_schedule(int iterations,int low,int high) {
        if(iterations<1 || low<1 || high<low || high>10) throw std::invalid_argument("invalid degree schedule");
        const int count=high-low+1,unit=count*(count+1)/2;
        std::vector<int> lengths(count);
        for(int k=0;k<count;++k) lengths[k]=(count-k)*(iterations/unit);
        // Exact port of Algo.h/DegreeScheduleDecreasingStages's remainder rule.
        int remaining=iterations%unit;
        for(int prefix=count;prefix>=1 && remaining>0;--prefix) {
            const int take=std::min(prefix,remaining);
            for(int k=0;k<take;++k) ++lengths[k];
            remaining-=take;
        }
        std::vector<int> out;
        out.reserve(iterations);
        for(int k=0;k<count;++k) out.insert(out.end(),lengths[k],low+k);
        return out;
    }
    Fit fit_analytic(const Matrix& y,const Statistics& stats,int degree,const AnalyticOptions& o) {
        validate_cloud(y,"analytic points");
        o.validate();
        if(stats.rho.size()!=y.rows() || stats.px.rows()!=y.rows() || stats.px.cols()!=y.cols() || !stats.rho.allFinite() || !stats.px.allFinite())
        throw std::invalid_argument("invalid analytic statistics");
        if(degree<o.min_degree || degree>o.max_degree) throw std::invalid_argument("requested degree outside options");
        std::vector<int> active;
        for(int i=0;i<y.rows();++i) {
            if(stats.rho[i]<0) throw std::invalid_argument("negative analytic fitting weight");
            if(stats.rho[i]>o.min_mass) active.push_back(i);
        }
        const int n=static_cast<int>(active.size()),d=static_cast<int>(y.cols());
        while(degree>o.min_degree && static_cast<int>(exponents(d,degree).size())>n) --degree;
        const Matrix phi=basis(y,degree);
        const int k=static_cast<int>(phi.cols());
        if(n<k) throw NumericalError("insufficient active rows for the minimum analytic degree");
        Matrix design(n,k),targets(n,d);
        double mass=0;
        for(int row=0;row<n;++row) {
            const int i=active[row];
            const double sw=std::sqrt(stats.rho[i]);
            mass+=stats.rho[i];
            design.row(row)=sw*phi.row(i);
            targets.row(row)=(sw/stats.rho[i])*stats.px.row(i);
        }
        // Paper Eq.(20)/(22): no hidden ridge penalty, coefficient projection,
        // displacement cap, or damping. SVD avoids explicitly forming Phi^T W Phi.
        require_finite(design,"weighted Taylor design");
        require_finite(targets,"weighted targets");
        // Rank-revealing complete orthogonal decomposition. Like the SVD it returns the
        // MINIMUM-NORM least-squares solution under rank deficiency and reports a rank
        // against the same relative threshold, so Eq.(20)/(22) is still solved without
        // regularization, damping or projection - only the factorization differs.
        // Measured on the real weighted Taylor designs: identical rank in every case and
        // fitted values agreeing to 4.6e-12, at 4.6x (degree 6) to 9.3x (degree 10) the
        // speed of BDCSVD. The coefficients themselves can differ more because the
        // monomial basis reaches condition 2e10 at degree 10; the algorithm consumes the
        // fitted values, and the stored map is evaluated on the same basis.
        Eigen::CompleteOrthogonalDecomposition<Matrix> solver(design);
        solver.setThreshold(o.rank_tolerance);
        Matrix absolute=solver.solve(targets);
        Matrix next=phi*absolute;
        require_finite(next,"analytic M-step");
        Matrix correction=absolute;
        for(int a=0;a<d;++a) correction(1+a,a)-=1;
        // exact representation A(p)=p+Phi(p)C
        double sse=0;
        for(int i:active) sse+=stats.rho[i]*(stats.px.row(i)/stats.rho[i]-next.row(i)).squaredNorm();
        return {
            AnalyticStep {
                degree,correction
            },next,n,static_cast<int>(solver.rank()),std::sqrt(sse/mass)
        };
    }
}
// namespace acpd

File: rust/crates/acpd-core/src/analytic.rs
//! Factorial-scaled Taylor mappings and exact, unregularized weighted fitting.
use crate::gaussian::Statistics;
use crate::types::*;
pub fn exponents(d: usize, degree: usize) -> RegResult<Vec<[usize;
3]>> {
    if ![2,3].contains(&d) || degree > 10 {
        return Err(invalid("basis requires dimension 2/3 and degree 0..10"));
    }
    let mut out = Vec::new();
    for r in 0..=degree {
        for a in (0..=r).rev() {
            if d == 2 {
                out.push([a,r-a,0]);
            }
            else {
                for b in (0..=r-a).rev() {
                    out.push([a,b,r-a-b]);
                }
            }
        }
    }
    Ok(out)
}
pub fn basis(points: &Matrix, degree: usize) -> RegResult<Matrix> {
    validate_cloud(points,"basis points")?;
    let powers = exponents(points.ncols(),degree)?;
    let phi = Matrix::from_fn(points.nrows(),powers.len(),|i,k| {
        let mut term = 1.0;
        for a in 0..points.ncols() {
            for j in 1..=powers[k][a] {
                term *= points[(i,a)]/j as f64;
            }
        }
        term
    });
    finite(&phi,"Taylor basis")?;
    Ok(phi)
}
/// Exact triangular-prefix remainder allocation from supplied Algo.h.
pub fn degree_schedule(iterations: usize, low: usize, high: usize) -> RegResult<Vec<usize>> {
    if iterations == 0 || low < 1 || high < low || high > 10 {
        return Err(invalid("invalid degree schedule"));
    }
    let count = high-low+1;
    let unit = count*(count+1)/2;
    let mut lengths: Vec<usize> = (0..count).map(|k| (count-k)*(iterations/unit)).collect();
    let mut remaining = iterations%unit;
    for prefix in (1..=count).rev() {
        let take = prefix.min(remaining);
        for length in lengths.iter_mut().take(take) {
            *length += 1;
        }
        remaining -= take;
        if remaining == 0 {
            break;
        }
    }
    let mut out = Vec::with_capacity(iterations);
    for (k,&length) in lengths.iter().enumerate() {
        out.extend(std::iter::repeat_n(low+k,length));
    }
    Ok(out)
}
#[derive(Debug, Clone)]
pub struct Fit {
    pub step: AnalyticStep, pub next: Matrix, pub active: usize, pub rank: usize, pub fit_rms: f64
}
pub fn fit_analytic(y: &Matrix, stats: &Statistics, mut degree: usize, o: &AnalyticOptions) -> RegResult<Fit> {
    validate_cloud(y,"analytic points")?;
    o.validate()?;
    if stats.rho.len() != y.nrows() || stats.px.shape() != y.shape()
    || !stats.rho.iter().chain(stats.px.iter()).all(|v| v.is_finite()) {
        return Err(invalid("invalid analytic statistics"));
    }
    if degree < o.min_degree || degree > o.max_degree {
        return Err(invalid("requested degree outside options"));
    }
    if stats.rho.iter().any(|v| *v < 0.0) {
        return Err(invalid("negative analytic fitting weight"));
    }
    let active: Vec<usize> = (0..y.nrows()).filter(|&i| stats.rho[i] > o.min_mass).collect();
    let n = active.len();
    let d = y.ncols();
    while degree > o.min_degree && exponents(d,degree)?.len() > n {
        degree -= 1;
    }
    let phi = basis(y,degree)?;
    let k = phi.ncols();
    if n < k {
        return Err(numerical("insufficient active rows for the minimum analytic degree"));
    }
    let design = Matrix::from_fn(n,k,|r,c| stats.rho[active[r]].sqrt()*phi[(active[r],c)]);
    let targets = Matrix::from_fn(n,d,|r,a| stats.px[(active[r],a)]/stats.rho[active[r]].sqrt());
    finite(&design,"weighted Taylor design")?;
    finite(&targets,"weighted targets")?;
    let svd = design.svd(true,true);
    let threshold = o.rank_tolerance*svd.singular_values.iter().copied().fold(0.0,f64::max);
    let rank = svd.singular_values.iter().filter(|&&s| s > threshold).count();
    // Minimum-norm ABSOLUTE mapping for a rank-deficient design, not an
    // identity-biased solution. No ridge, projection, cap, or damping.
    let absolute = svd.solve(&targets,threshold).map_err(numerical)?;
    let next = &phi*&absolute;
    finite(&next,"analytic M-step")?;
    let mut coefficients = absolute;
    for a in 0..d {
        coefficients[(1+a,a)] -= 1.0;
    }
    let mass: f64 = active.iter().map(|&i| stats.rho[i]).sum();
    let sse: f64 = active.iter().map(|&i| stats.rho[i]*(0..d).map(|a| (stats.px[(i,a)]/stats.rho[i]-next[(i,a)]).powi(2)).sum::<f64>()).sum();
    Ok(Fit {
        step:AnalyticStep {
            degree,coefficients
        },next,active:n,rank,fit_rms:(sse/mass).sqrt()
    })
}
